src/persist/benches/consensus.rs

```diff
@@ -0,0 +1,95 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Benchmarks for different persistent Write implementations.
+
+use std::path::Path;
+use std::time::{Duration, Instant};
+
+use criterion::measurement::WallTime;
+use criterion::{Bencher, BenchmarkGroup, BenchmarkId, Throughput};
+
+use mz_persist::location::{Consensus, SeqNo, VersionedData};
+use mz_persist::mem::MemConsensus;
+use mz_persist::sqlite::SqliteConsensus;
+use mz_persist::workload::{self, DataGenerator};
+
+fn new_sqlite_consensus(path: &Path) -> SqliteConsensus {
+    SqliteConsensus::open(&path).expect("creating a SqliteConsensus cannot fail")
+}
+
+// Benchmark the write throughput of Consensus::compare_and_set.
+fn bench_compare_and_set<C: Consensus>(
+    consensus: &mut C,
+    key: &str,
+    data: Vec<u8>,
+    b: &mut Bencher,
+) {
+    let deadline = Instant::now() + Duration::from_secs(3600);
+
+    // Set an initial value in case that has different performance characteristics
+    // for some Consensus implementations.
+    //
+    // The benchmark runner might execute this code several times to warm-up, grab
+    // the existing value if one has already been set.
+    let mut current = match futures_executor::block_on(consensus.compare_and_set(
+        key,
+        deadline,
+        None,
+        VersionedData {
+            seqno: SeqNo(0),
+            data: data.clone(),
+        },
+    ))
+    .expect("gave invalid inputs")
+    {
```

Comment:
make this block an expect also, with a message something like "initial value unexpectedly set"

(then instead of setting the initial value of current by returning from this, you could do something like `let mut current = SeqNo::minimum(); consensus.compare_and_set(..., Some(current), ...)` similar to how it's called below)

Reply:
I don't think we can do this -- I tried to explain the rationale in the comment above but criterion runs the benchmark several times as a warmup so the initial value gets set. We get around this a different way in the blob writer benchmarks by picking a random key I can do that instead if thats better, and then we can start with SeqNo(0) each time?

Reply:
yeah i think we should use a random key to make the runs as identical as possible